src/diagram/workflow_builder.rs

```diff
@@ -0,0 +1,501 @@
+            }
+            DiagramOperation::Split(split_op) => {
+                let next_op_ids: Vec<&OperationId> = match &split_op.params {
+                    SplitOpParams::Index(v) => v.iter().collect(),
+                    SplitOpParams::Key(v) => v.values().collect(),
+                };
+                for next_op_id in next_op_ids {
+                    add_edge(op_id, next_op_id, EdgeState::Pending)?;
+                }
+                if let Some(remaining) = &split_op.remaining {
+                    add_edge(op_id, &remaining, EdgeState::Pending)?;
+                }
+            }
+            DiagramOperation::Join(join_op) => {
+                add_edge(op_id, &join_op.next, EdgeState::Pending)?;
+            }
+            DiagramOperation::Transform(transform_op) => {
+                add_edge(op_id, &transform_op.next, EdgeState::Pending)?;
+            }
+            DiagramOperation::Dispose => {}
+        }
+    }
+
+    let mut unconnected_vertices: Vec<&Vertex> = vertices.values().collect();
+    while unconnected_vertices.len() > 0 {
+        let ws = unconnected_vertices.clone();
+        let ws_length = ws.len();
+        unconnected_vertices.clear();
+
+        for v in ws {
+            let in_edges: Vec<&Edge> = v.in_edges.iter().map(|idx| &edges[idx]).collect();
+            if in_edges
+                .iter()
+                .any(|e| matches!(e.state, EdgeState::Pending))
+            {
+                // not all inputs are ready
+                debug!(
+                    "defer connecting [{}] until all incoming edges are ready",
+                    v.op_id
+                );
+                unconnected_vertices.push(v);
+                continue;
+            }
+
+            connect_vertex(builder, registry, &vertices, &mut edges, &inputs, v)?;
+        }
+
+        // can't connect anything and there are still remaining vertices
+        if unconnected_vertices.len() > 0 && ws_length == unconnected_vertices.len() {
+            warn!(
+                "the following operations are not connected {:?}",
+                unconnected_vertices
+                    .iter()
+                    .map(|v| v.op_id)
+                    .collect::<Vec<_>>()
+            );
+            return Err(DiagramError::BadInterconnectChain);
+        }
+    }
+
+    Ok(())
+}
+
+fn connect_vertex<'a>(
+    builder: &mut Builder,
+    registry: &NodeRegistry,
+    vertices: &'a HashMap<&OperationId, Vertex>,
+    edges: &mut HashMap<usize, Edge<'a>>,
+    inputs: &HashMap<&OperationId, DynInputSlot>,
+    target: &'a Vertex,
+) -> Result<(), DiagramError> {
+    debug!("connecting [{}]", target.op_id);
+    match target.op {
+        // join needs all incoming edges to be connected at once so it is done at the vertex level
+        // instead of per edge level.
+        DiagramOperation::Join(_) => {
+            if target.in_edges.is_empty() {
+                if target.in_edges.is_empty() {
```

Comment:
This seems redundant with the previous check

Reply:
08631b4c15a8dded54452e4783d0bfb80dea83f0 oops, copy paste error